### PR TITLE
[docs] Add architecture doc.

### PR DESCRIPTION
**Pull request by silvasean**
This attempts to get out of my head most of the critical layering and
project structure decisions for Torch-MLIR.

### CODE REVIEW

**Review by powderluv — APPROVED** [review]
Wow much needed. I have some thoughts but will discuss in the next call.

**Comment by silvasean** [comment]
> Wow much needed. I have some thoughts but will discuss in the next call.

Hi Anush, that is not very inclusive to the community. We should try to keep our discussion in the open where possible.

**Comment by silvasean** [comment]
> Wow much needed. I have some thoughts but will discuss in the next call.

Sorry, I interpreted this as one of our other calls, but I now realize that you probably meant the developer hour next week. Apologies for the mistaken intepretation.

**Review by henrytwo — APPROVED** [review]
LGTM! Thanks for putting this together -- this should make it much easier for new people to get up to speed with the project

File: docs/architecture.md [hunk 1]
```diff
@@ -0,0 +1,438 @@
...
+The Torch-MLIR project provides core infrastructure for bridging between the
+Torch ecosystem and the MLIR ecosystem. Torch-MLIR does not attempt to provide a
+production end-to-end flow for PyTorch programs by itself, but is a useful
+component for constructing one.
+
+## Overview
+
+Torch-MLIR has two parts, which we call the "frontend" and "backend". These two
+halves interface at an abstraction layer which we call the "backend contract"
+and which is a normalized subset of the `torch` dialect.
```
Comment (sjain-stanford) [inline docs/architecture.md:14]: If it's not too much work, it might help to accompany this with figures (like the one on the README).

File: docs/architecture.md [hunk 2]
```diff
@@ -0,0 +1,438 @@
...
+
+The frontend of Torch-MLIR is concerned with interfacing to PyTorch itself, and
+then normalizing the program to the "backend contract". This part involves build
+system complexity and exposure to PyTorch API's to get the program into the MLIR
+`torch` dialect in the first place. When we interface with TorchScript, we
+additionally have a large amount of lowering and simplification to do within
+MLIR on the `torch` dialect.
```
Comment (sjain-stanford) [inline docs/architecture.md:21]: Is my understanding right that the frontend will change more frequently (to track changes in PyTorch ecosystem such as TorchDynamo), however the backend contract (middle-end?) would isolate / shield the lowerings to the respective backends from being impacted too much or at all? If that is correct, maybe include a statement to clarify this?

File: docs/architecture.md [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+This dialect supports progressive lowering from the raw imported PyTorch
+programs that various PyTorch integration points provide, all the way down to
+the backend contract.
+
+The `torch` dialect must be versatile enough to support being imported by any
+program capture mechanism in PyTorch -- this could be TorchDynamo, `torch.fx`,
+LazyTensorCore, TorchScript, `torch.jit.trace`, etc. Thankfully, PyTorch is
+factored such that we can handle this with one core import path, which is
+through the PyTorch "[JIT
+IR](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md)",
```
Comment (sjain-stanford) [inline docs/architecture.md:52]: nit: Maybe better to use a permalink?

File: docs/architecture.md [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
+### Ops
+
+See [TorchOps.td](https://github.com/llvm/torch-mlir/blob/114f48e96c578ee76a6f83b3aa4aa229a8d5b76e/include/torch-mlir/Dialect/Torch/IR/TorchOps.td#L1)
+
+The ops in the `torch` dialect are almost entirely generated based on the
+PyTorch JIT IR operator registry via the script
+[torch_ods_gen.py](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/python/torch_mlir/dialects/torch/importer/jit_ir/build_tools/torch_ods_gen.py#L1) (invoked via [update_torch_ods.sh](https://github.com/llvm/torch-mlir/blob/main/build_tools/update_torch_ods.sh)).
+This script queries the registry and generates MLIR
+[ODS](https://mlir.llvm.org/docs/OpDefinitions/) in
+[GeneratedTorchOps.td](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/include/torch-mlir/Dialect/Torch/IR/GeneratedTorchOps.td#L1). We have a guide for [adding a new op end-to-end](https://github.com/llvm/torch-mlir/wiki/Torch-ops-E2E-implementation).
```
Comment (sjain-stanford) [inline docs/architecture.md:72]: (can be addressed as a follow-on) Would it help to move this wiki page (and any other wiki page) as a Markdown doc tracked in code?

File: docs/architecture.md [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+
+- All tensors have been converted to value semantics.
+- All tensors have at least a known number of dimensions (i.e. rank), and
+  ideally also have a precise size for each dimension.
+- All tensors have a known dtype.
+- Certain ops have been decomposed to make them easier to handle (this is
+  configurable).
+
+See the extensive comments in the function `satisfiesBackendContract` (and its
+callees) in the `LowerToBackendContract` pass for an extended rationale for
```
Comment (sjain-stanford) [inline docs/architecture.md:127]: (at the risk of making the splits even more confusing) Does it make sense to use "middle-end" instead of "backend-contract"? 

Of course this would be a change in many other places than the docs (so not to be considered for this PR), but just wanted to put it out in the context of this architecture doc.

File: docs/architecture.md [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+certain highly stable enums and related logic must be shared with upstream
+PyTorch, and for those we copy code from PyTorch into
+[TorchUpstream.h](https://github.com/llvm/torch-mlir/blob/fde390c7669e29362b18388448ef2b188713383f/include/torch-mlir/Dialect/Torch/Utils/TorchUpstream.h#L13).
+
+### LazyTensorCore
+
+Docs: https://github.com/llvm/torch-mlir/blob/main/docs/ltc_backend.md
+
+LazyTensorCore (LTC) is a program capture method provided by PyTorch that does
+device-level tracing. This low-level intercept point sits below gradient
```
Comment (sjain-stanford) [inline docs/architecture.md:163]: nit: interception?

File: docs/architecture.md [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+   - Shape refinement. See [shape_lib.md](https://github.com/llvm/torch-mlir/blob/main/docs/shape_lib.md) for detail
+   - DType refinement. See `RefineTypes`.
+   - Decomposing ops into more primitive ops. See `DecomposeComplexOps`.
+
+## The Backends
+
+Torch-MLIR provides 3 built-in backends, which take the backend contract IR and
+lower it to the requirements of each backend. The 3 backends are:
+
+- Linalg-on-Tensors (+ `arith`, `tensor`, etc.)
```
Comment (sjain-stanford) [inline docs/architecture.md:265]: Is there a link to this dialect we could embed here?

File: docs/architecture.md [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+which provide a default conversion from `torch` dialect types to the builtin
+`tensor` type and scalar integer/float types. These are not the right choice for
+all backends, but can be copied and adapted by backends. These files closely
+follow the "Type Conversions the Not-So-Hard Way" talk.
+
+
+## Testing
+
+See
+[development.md](https://github.com/llvm/torch-mlir/blob/main/development.md#testing)
```
Comment (sjain-stanford) [inline docs/architecture.md:341]: nit: permalink?

**Comment by powderluv** [comment]
Apologies for the brevity was putting my son down to sleep and had to swipe type 😀

Anyway I was wondering how we want to capture upcoming architectural / roadmap discussions. We can open an RFC issue for it or track some other way.

Things I would be interested in hearing feedback on:

* Op authoring and custom op support
* How do we support things that are not native to Pytorch yet (e.g: custom numerics, sparsity)
* How do we want to version our ops ? (Any learnings from Onnx op set version or Pytorch native_functions.yml )?
* Any thoughts on stable branches and release? (Anyone getting to that point ?)

None of these have to be addressed in this PR and I'm sure there are more. Maybe a high level Roadmap.md can be a catch all ?

File: docs/architecture.md [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+// CHECK: bar
+```
+
+Such a test is simply checking that the implementation of an op is the way it
+is. There is no way to change the implementation while having the test pass. So
+the test is fully redundant with the implementation.
+
+Because of this, many Torch-MLIR patches adding support for new ops have no
+`.mlir` unit tests, and only include an end-to-end test (or more likely,
+multiple to test various facets of the op). We generally make sure that our
```
Comment (sjain-stanford) [inline docs/architecture.md:391]: nit: `multiple, `

Or, maybe reword this as: 
`only include end-to-end test(s)`

**Review by sjain-stanford — APPROVED** [review]
Thank you for writing this up. As a newcomer to this codebase, this helps my mental model quite a bit. Left some non-blocking nits / comments.

**Comment by sjain-stanford** [comment]
Should we link `docs/architecture.md` from the main `README.md`?

File: docs/architecture.md [hunk 10]
```diff
@@ -0,0 +1,438 @@
+# Torch-MLIR Architecture
+
+## Introduction
+
+The Torch-MLIR project provides core infrastructure for bridging between the
+Torch ecosystem and the MLIR ecosystem. Torch-MLIR does not attempt to provide a
```
[the remark this answers is not in the document]
Comment (silvasean) [reply docs/architecture.md:6]: Good point. I think keeping the high level as "bridging" gives us the scope freedom. I don't want to pigeonhole the project scope too much.

Added your suggested sentence too.

File: docs/architecture.md (hunk 1 above)
[the remark this answers is not in the document]
Comment (silvasean) [reply docs/architecture.md:14]: Normalized in this context means that it is made to obey certain norms. I rewrote it as "These two
halves interface at an abstraction layer that we call the "backend contract",
which is a subset of the `torch` dialect with certain properties appealing for
backends to lower from."

File: docs/architecture.md [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+## Overview
+
+Torch-MLIR has two parts, which we call the "frontend" and "backend". These two
+halves interface at an abstraction layer which we call the "backend contract"
+and which is a normalized subset of the `torch` dialect.
+
+The frontend of Torch-MLIR is concerned with interfacing to PyTorch itself, and
+then normalizing the program to the "backend contract". This part involves build
+system complexity and exposure to PyTorch API's to get the program into the MLIR
+`torch` dialect in the first place. When we interface with TorchScript, we
```
[the remark this answers is not in the document]
Comment (silvasean) [reply docs/architecture.md:19]: It is just to TorchScript. That is what I was aiming for with saying "additionally"

File: docs/architecture.md (hunk 2 above)
Comment (silvasean) [reply docs/architecture.md:21]: At this point I'm not really able to make a statement about the stability of the backend contract. In some ways it can be less stable than the frontends, at least today. For example, if PyTorch changes an op or starts using a new op in place of an old op, that will, at least today, end up immediately reflected in our backend contract. The actual code for interfacing to PyTorch doesn't need to change for that to happen.

File: docs/architecture.md (hunk 3 above)
Comment (silvasean) [reply docs/architecture.md:52]: Good idea.

File: docs/architecture.md [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+the backend contract.
+
+The `torch` dialect must be versatile enough to support being imported by any
+program capture mechanism in PyTorch -- this could be TorchDynamo, `torch.fx`,
+LazyTensorCore, TorchScript, `torch.jit.trace`, etc. Thankfully, PyTorch is
+factored such that we can handle this with one core import path, which is
+through the PyTorch "[JIT
+IR](https://github.com/pytorch/pytorch/blob/master/torch/csrc/jit/OVERVIEW.md)",
+and lives in
+[torch-mlir/python/torch_mlir/dialects/torch/importer/jit_ir](https://github.com/llvm/torch-mlir/tree/e322f6a8784009b37aa354abfa9a40a80f30877d/python/torch_mlir/dialects/torch/importer/jit_ir).
```
[the remark this answers is not in the document]
Comment (silvasean) [reply docs/architecture.md:54]: I am using permalinks to make the document a bit more stable to us moving directories and stuff.

File: docs/architecture.md (hunk 4 above)
Comment (silvasean) [reply docs/architecture.md:72]: Yes, I've been thinking the same thing.

Comment (silvasean) [reply docs/architecture.md:72]: This dates to a much earlier phase of the project where we didn't have the same docs structure.

File: docs/architecture.md [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+progressive lowering from raw imported IR which maybe be missing shapes, dtypes,
+and value semantics, into the backend contract which provides those. Torch-MLIR
+has two tensor types `ValueTensorType` (`!torch.vtensor`) and
+`NonValueTensorType` (`!torch.tensor`) sharing most of their definition in
+[TorchTypes.td](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/include/torch-mlir/Dialect/Torch/IR/TorchTypes.td#L58).
+These types have a common C++ base class
+[`BaseTensorType`](https://github.com/llvm/torch-mlir/blob/e322f6a8784009b37aa354abfa9a40a80f30877d/include/torch-mlir/Dialect/Torch/IR/TorchTypes.h#L40)
+which permits abstracting across them. Both `ValueTensorType` and
+`NonValueTensorType` have an optional list of optional sizes and an optional
+dtype.
```
[the remark this answers is not in the document]
Comment (silvasean) [reply docs/architecture.md:108]: I added a little bit extra talking about it. Is there any particular point of clarification you would like besides pointing out that one is aliasing/mutable and the other is not?

File: docs/architecture.md (hunk 5 above)
Comment (silvasean) [reply docs/architecture.md:127]: I would consider a "middle end" as a place where we do a large number of transformations within a common IR form. Today, the backend contract is just an interface and we don't do much on it. I think if we grow to do more transformations at this level of abstraction, then we can rename it to "middle end".

File: docs/architecture.md [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+This bypasses a number of very complex dependency and shared library issues.
+
+Additionally, we maintain the invariant that the core MLIR compiler code (in
+`lib/` and `include/`) never has a build dependency on PyTorch itself. This
+strict isolation avoids a number of complex dependency issues and ensures that
+`torch-mlir-opt` and similar debugging tools always provide the excellent
+development and debugging experience that MLIR developers expect. Sometimes,
+certain highly stable enums and related logic must be shared with upstream
+PyTorch, and for those we copy code from PyTorch into
+[TorchUpstream.h](https://github.com/llvm/torch-mlir/blob/fde390c7669e29362b18388448ef2b188713383f/include/torch-mlir/Dialect/Torch/Utils/TorchUpstream.h#L13).
```
[the remark this answers is not in the document]
Comment (silvasean) [reply docs/architecture.md:156]: Good point. Moved later.

**Comment by silvasean** [comment]
> Apologies for the brevity was putting my son down to sleep and had to swipe type grinning
> 
> Anyway I was wondering how we want to capture upcoming architectural / roadmap discussions. We can open an RFC issue for it or track some other way.
> 
> Things I would be interested in hearing feedback on:
> 
> * Op authoring and custom op support
> * How do we support things that are not native to Pytorch yet (e.g: custom numerics, sparsity)
> * How do we want to version our ops ? (Any learnings from Onnx op set version or Pytorch native_functions.yml )?
> * Any thoughts on stable branches and release? (Anyone getting to that point ?)
> 
> None of these have to be addressed in this PR and I'm sure there are more. Maybe a high level Roadmap.md can be a catch all ?

These are all really good points, I definitely think they are more forward looking than this doc, which attempts to capture the present state. roadmap.md or evolution.md could be good candidates.

**Comment by sstamenova** [comment]
This is really good, thank you! I'll abstain from making any nitpicks on apostrophes :)
